### PR TITLE
RFC: migrate from Docopt to StructOpt

### PR DESCRIPTION
**Pull request by chriskrycho**
Use [StructOpt] to get automatic coloration, completions, and (in the author's experience) easier development and maintenance.

[StructOpt]: https://docs.rs/structopt/0.2.14/structopt/

[Rendered](https://github.com/chriskrycho/notion-rfcs/blob/cli-ui/text/0000-cli-ui.md)

### CODE REVIEW

**Comment by chriskrycho** [comment]
Happily, no, because we won't actually be using the `template` field at this point – if it ends up being useable for us later, it would be if we made e.g. `ansi_term` support `const fn` (if that's possible; it may not be!) and thus got its cross-platform support baked in for free. The basic colorization stuff all Just Works cross-platform out of the box.

File: text/0000-cli-ui.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+- The configuration is validated by the compiler rather than at runtime, and the compiler’s error messages are usually much clearer about what is wrong than are the runtime errors from Doctopt.
+
+- It properly passes through global options to subcommands, and it understands all variants of “help” invocations (position and both flag styles) out of the box, and can be trivially configured to display help/usage information when no arguments are passed.
+
+Though StructOpt is not *universally* better than Docopt, it can generate similar output, but with a lower maintenance burden and a number of niceties *not* supported by Docopt.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+This is purely an internal, developer-facing change, with *benefits* to the user but no changes in actual use of Notion. The documentation for StructOpt (and its core dependency [`clap`]) are sufficient for developers working on the project.
```
Comment (dherman) [inline text/0000-cli-ui.md]: I find this sentence confusing: it sounds like it's saying "this is user-visible but it's invisible to the user." Do you mean it does have user benefits but doesn't have any changes to workflow?

Comment (chriskrycho) [reply text/0000-cli-ui.md]: Ah, great point. I'll update to that effect; thank you!

File: text/0000-cli-ui.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
+- Feature Name: cli-ui-improvements
+- Start Date: 2019-02-25
+- RFC PR: (leave this empty)
+- Notion Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Transition from [Docopt] to [StructOpt], for an easier path to nice user experience (and better developer experience when working on) the command line interface.
```
Comment (dherman) [inline text/0000-cli-ui.md]: Are completions the only user-visible change? Is there an implication for color? You mention it later but it's not clear to me if DocOpt doesn't support color and StructOpt does?

**Comment by dherman** [comment]
I'm 👍 on this. I like DocOpt but agree that the runtime parsing, unhelpful errors, and lack of documentation around the input grammar are problematic. And generally I think it's not super actively maintained so that's probably another good reason to switch.

File: text/0000-cli-ui.md (hunk 2 above)
Comment (chriskrycho) [reply text/0000-cli-ui.md]: Ah, good clarification. Docopt *can* support them, but you have to manually do every bit of it with e.g. `format!` and `ansi_term`; they work out of the box (and are pretty nice!) with StructOpt.

**Comment by chriskrycho** [comment]
Thanks, @dherman and @charlespierce – I've updated the RFC to clarify those points!

**Comment by dherman** [comment]
Discussion on this RFC has stabilized and the implementation has already landed and is working out. I'm marking this as final comment period and we can probably merge soon.

**Pull request merged**